text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,325 @@
+> better, and provide a more streamlined user-facing API,
+> both while staying backwards compatible.
+
+## Working Backwards
+
+The final API for CDK Pipelines is finally here. Compared to the previous API:
+
+- CDK applications could not be deployed in waves. They can now, as we automatically establish runorders between
+  actions.
+- You could not build pipelines that used multiple sources to feed into the synth, or that had a custom build step. You can now.
+- Large CDK applications (> 25 stacks) could not previously be deployed because they would exceed the 50 actions per
+  stage limit that CodePipeline has. With the new implementation, the actions will automatically spread over multiple
+  stages, even if you deploy multiple applications in parallel.
+- For simple use cases, you no longer need to manage CodePipeline Artifacts: artifact management is implicit
+  as objects that produce artifacts can be used to reference them.
+
+Best of all: your old code--even though it uses no longer recommended idioms--still continues to work, so you
+can upgrade to the new style at your own pace.
+
+The following:
+
+```ts
+const sourceArtifact = new codepipeline.Artifact();
+const cloudAssemblyArtifact = new codepipeline.Artifact('CloudAsm');
+const integTestArtifact = new codepipeline.Artifact('IntegTests');
+
+const pipeline = new cdkp.CdkPipeline(this, 'Pipeline', {
+  cloudAssemblyArtifact,
+
+  // Where the source can be found
+  sourceAction: new codepipeline_actions.GitHubSourceAction({
+    actionName: 'GitHub',
+    output: sourceArtifact,
+    oauthToken: SecretValue.secretsManager('github-token'),
+    owner: 'OWNER'
+    repo: 'REPO',
+    trigger: codepipeline_actions.GitHubTrigger.POLL,
+  }),
+
+  // How it will be built
+  synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
+    sourceArtifact,
+    cloudAssemblyArtifact,
+    projectName: 'MyServicePipeline-synth',
+    additionalArtifacts: [
+      {
+        directory: 'test',
+        artifact: integTestArtifact,
+      },
+    ],
+  }),
+});
+
+const stage = pipeline.addApplicationStage(new MyStage(this, 'PreProd', {
+  env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
+}));
+stage.addActions(
+  new cdkp.ShellScriptAction({
+    actionName: 'UseSource',
+    commands: [
+      // Comes from source
+      'cat README.md',
+    ],
+    additionalArtifacts: [sourceArtifact],
+  }),
+);
+```
+
+Becomes:
+
+```ts
+const source = cdkp.Source.gitHub('OWNER/REPO');
+
+const pipeline = new cdkp.CdkPipeline(this, 'Pipeline', {
+  synthAction: cdkp.Build.standardNpmSynth({
+    input: source,
+    projectName: 'MyServicePipeline-synth',
+    additionalOutputs: [
+      { directory: 'test', name: 'tests' },
+    ],
+  }),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
```

Comment:
What does it mean to use `CDK_DEFAULT_XXX` here? Which environment will this use in the Pipeline?

Reply:
Maybe a poorly chosen example. I copy/pasted this from integ tests.